engine/src/main/java/org/pentaho/di/trans/steps/ivwloader/IngresVectorwiseLoader.java

```diff
@@ -0,0 +1,995 @@
+        fileOutputStream.close();
+      }
+    }
+  }
+
+  static class SqlRunner extends Thread {
+    private IngresVectorwiseLoaderData data;
+
+    private String loadCommand;
+
+    private Exception ex;
+
+    SqlRunner( IngresVectorwiseLoaderData data, String loadCommand ) {
+      this.data = data;
+      this.loadCommand = loadCommand;
+    }
+
+    public void run() {
+      try {
+        data.sqlOutputStream.write( data.getBytes( loadCommand ) );
+        data.sqlOutputStream.flush();
+      } catch ( Exception ex ) {
+        this.ex = ex;
+      }
+    }
+
+    void checkExcn() throws Exception {
+      // This is called from the main thread context to re-throw any saved
+      // exception.
+      //
+      if ( ex != null ) {
+        throw ex;
+      }
+    }
+  }
+
+  public class VWloadMonitor implements Runnable {
+    private Process vwloadProcess;
+    private LogWriter logWriter;
+    private Thread outputLoggerThread;
+
+    VWloadMonitor( Process loadProcess, LogWriter logWriter, Thread outputLoggerThread ) {
+      this.vwloadProcess = loadProcess;
+      this.logWriter = logWriter;
+      this.outputLoggerThread = outputLoggerThread;
+    }
+
+    public void run() {
+      try {
+        int resultCode = vwloadProcess.waitFor();
+        Long[] results = tryToParseVWloadResultMessage();
+        if ( results != null ) {
+          setLinesOutput( results[1] );
+          setLinesRejected( results[2] );
+        }
+        boolean errorResult =
+          ( resultCode != 0 ) || ( results != null && ( !meta.isContinueOnError() && !meta.isUsingVwload() ) );
+        if ( errorResult ) {
+          setLinesOutput( 0L );
+          logError( "Bulk loader finish unsuccessfully" );
+          setErrors( 1L );
+        } else {
+          setErrors( 0L );
+        }
+      } catch ( Exception ex ) {
+        setErrors( 1L );
+        logError( "Unexpected error encountered while monitoring bulk load process", ex );
+      }
+    }
+
+    @SuppressWarnings( "resource" )
+    private Long[] tryToParseVWloadResultMessage() throws InterruptedException, IOException {
+      outputLoggerThread.join();
+      Long[] result = new Long[3];
+      if ( meta.isUsingVwload() ) {
+        String lastLine = logWriter.getLastInputStreamLine();
+        Scanner sc = null;
+        try {
+          sc = new Scanner( lastLine );
+          sc = sc.useDelimiter( "\\D+" );
+          int i = 0;
+          while ( sc.hasNext() ) {
+            result[i++] = sc.nextBigInteger().longValue();
+          }
+        } finally {
+          if ( sc != null ) {
+            sc.close();
+          }
+        }
+      } else {
+        if ( meta.getErrorFileName() == null ) {
+          return null;
+        }
+        File errorFile = new File( meta.getErrorFileName() );
+        if ( !errorFile.exists() ) {
+          return null;
+        } else {
+          LineNumberReader lnr = new LineNumberReader( new FileReader( errorFile ) );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *lnr*. Execution paths that do not contain closure statements were detected. Ensure *lnr* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).